File: src/sites/dreamcamvr.cpp
// ─────────────────────────────────────────────────────────────────
// DreamCamVR site plugin — VR extension of DreamCam
// Port from Python: uses "video3D" streamType instead of "video2D"
// VR format suffix from stream URL query parameters
// ─────────────────────────────────────────────────────────────────

#include "sites/dreamcamvr.h"

namespace sm
{

    REGISTER_SITE(DreamCamVR);

    const std::map<std::string, std::string> DreamCamVR::kVrFrameFormatMap = {
        {"FISHEYE", "F"},
        {"PANORAMIC", "P"},
        {"CIRCULAR", "C"}};

    DreamCamVR::DreamCamVR(const std::string &username)
        : DreamCam(kSiteName, kSiteSlug, username)
    {
        streamType_ = "video3D"; // Override to look for 3D streams
    }

} // namespace sm

File: streamonitor/sites/dreamcam_vr.py
from urllib.parse import urlparse, parse_qs
from typing import Tuple, List

from parameters import VR_FORMAT_SUFFIX
from streamonitor.bot import Bot

from streamonitor.downloaders.fmp4s_wss import getVideoWSSVR
from streamonitor.sites.dreamcam import DreamCam


class DreamCamVR(DreamCam):
    site = 'DreamCamVR'
    siteslug = 'DCVR'

    _stream_type = 'video3D'

    def get_site_color(self) -> Tuple[str, List[str]]:
        """Return the color scheme for this VR site."""
        return ("light_cyan", ["bold"])

    vr_frame_format_map = {
        'FISHEYE': 'F',
        'PANORAMIC': 'P',
        'CIRCULAR': 'C',
    }

    def __init__(self, username):
        super().__init__(username)
        self.getVideo = getVideoWSSVR
        self.stopDownloadFlag = False

    @property
    def filename_extra_suffix(self):
        vr_suffix = ''
        if not VR_FORMAT_SUFFIX:
            return vr_suffix

        video_url = self.getVideoUrl()
        query = parse_qs(urlparse(video_url).query)
        if 'stereoPacking' in query:
            vr_packing = query["stereoPacking"][0]
            vr_suffix += f'_{vr_packing}'
        if 'frameFormat' in query and "horizontalAngle" in query:
            vr_frame_format = self.vr_frame_format_map[query["frameFormat"][0]]
            vr_angle = query["horizontalAngle"][0]
            vr_suffix += f'_{vr_frame_format}{vr_angle}'
        return vr_suffix

    
    def isMobile(self):
        return False
